### DISCUSSION TITLE
Separate classes for Validation Rules

### DISCUSSION THREAD

**Discussion by pxlrbt** [laravel/framework#49256] [Ideas]
Hey,

Laravel and PHP get safer to use with every release. We have Enums and use classes a lot more for configuration then strings nowadays. But we still have this set up strings to define Validation Rules. Extra arguments are chained onto the string which makes it even worse. 

I'd love to see rules move to their own classes. This gives us auto-completion, less typos and methods to configure the rules. Using the example from the docs, we could transform this:

```php
$request->validate([
  'title' => 'required|unique:posts|max:255',
  'body' => 'required',
]);
```

to this:

```php
use \Illuminate\Validation\Rules;

$request->validate([
  'title' => [
    new Rules\Required,
    new Rules\Unique(table: 'posts'),
    new Rules\Max(limit: 255)
  ],
  'body' => new Rules\Required,
]);
```

What are your thoughts on this?

**Comment by shaedrich** [comment]
It already is possible in many, but not all, cases:

via [`Illuminate\Validation\Rule::*`](https://laravel.com/api/10.x/Illuminate/Validation/Rule.html):
* `unique`
* `exists`
* `in`
* `notIn`
* `requiredIf`
* `excludeIf`
* `prohibitedIf`
* `enum`

via [`Illuminate\Validation\Rules\*`](https://laravel.com/api/10.x/Illuminate/Validation/Rules.html):
* [`Can`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/Can.html)
* [`DatabaseRule`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/DatabaseRule.html)
* [`Dimensions`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/Dimensions.html)
* [`Enum`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/Enum.html)
* [`ExcludeIf`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/ExcludeIf.html)
* [`Exists`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/Exists.html)
* [`File`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/File.html)
* [`ImageFile`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/ImageFile.html)
* [`In`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/In.html)
* [`NotIn`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/NotIn.html)
* [`Password`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/Password.html)
* [`ProhibitedIf`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/ProhibitedIf.html)
* [`RequiredIf`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/RequiredIf.html)
* [`Unique`](https://laravel.com/api/10.x/Illuminate/Validation/Rules/Unique.html)

But yeah, it would be easier if one didn't have to look up whether a certain rule already has a class or not.

**Reply by pxlrbt** [reply]
Yeah, figured that out as well. So should be acceptable to PR the missing classes I guess.

**Reply by shaedrich** [reply]
Acceptable, sure. If Taylor is in the mood to accept it, who knows 🤷🏻
